R/utils.R

```diff
@@ -11,132 +9,112 @@
 # `write` operation) to be undone, AND for any future `writeWS` calls to be
 # ignored. A call to `write` will be respected, and will restore normal
 # behavior.
-WSTextWriter <- R6Class("WSTextWriter",
-  private = list(
-    writer = "TextWriter",
-    suppressing = logical(1)
-  ),
-  public = list(
-    initialize = function() {
-      private$writer <- TextWriter$new()
-      private$suppressing <- FALSE
-    },
-    close = function() {
-      private$writer$close()
-    },
-    write = function(text) {
-      private$writer$write(text)
+#
+# Text is automatically converted to UTF-8 before being written.
+#' @param bufferSize The initial size of the buffer in which writes are stored.
+#'   The buffer will be periodically cleared, if possible, to cache the writes
+#'   as a string. If the buffer cannot be cleared (because of the need to be
+#'   able to backtrack to fulfill an `eatWS()` call), then the buffer size will
+#'   be doubled.
+#' @noRd
+WSTextWriter <- function(bufferSize=1024){
+  # A string into which we can collapse entries in the buffer if the buffer gets full.
+  accumulated <- ""
```

Comment:
Instead of having accumulated, try out having just collapsing into the first element in the buffer.

Make the first element in the array the collapsed writes that are persisted.
Make the second element in the array everything written since the last marked write position (all the whitespace).